prov/gni/src/gnix_xpmem.c

```diff
@@ -0,0 +1,603 @@
+/*
+ * Copyright (c) 2016 Los Alamos National Security, LLC.
+ *                    All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        opyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if HAVE_CONFIG_H
+#include <config.h>
+#endif /* HAVE_CONFIG_H */
+
+#include "gnix.h"
+#include "gnix_mr.h"
+#include "gnix_hashtable.h"
+#include "gnix_xpmem.h"
+
+
+#if HAVE_XPMEM
+
+bool gnix_xpmem_disabled = false;
+
+#define XPMEM_PAGE_SIZE 4096
+
+static  pthread_mutex_t gnix_xpmem_lock = PTHREAD_MUTEX_INITIALIZER;
+static xpmem_segid_t gnix_my_segid;
+static int gnix_xpmem_ref_cnt;
+
+static void *__gnix_xpmem_attach_seg(void *handle,
+				     void *address,
+				     size_t length,
+				     struct _gnix_fi_reg_context *,
+				     void *context);
+
+static int __gnix_xpmem_detach_seg(void *handle,
+				   void *context);
+
+static int __gnix_xpmem_destroy_mr_cache(void *context);
+
+struct gnix_xpmem_ht_entry {
+	struct gnix_mr_cache *mr_cache;
+	struct gnix_xpmem_handle *xp_hndl;
+	xpmem_apid_t apid;
+};
+
+/*
+ * TODO: should be adjustable from domain params
+ */
+static gnix_mr_cache_attr_t _gnix_xpmem_default_mr_cache_attr = {
+		.soft_reg_limit      = 128,
+		.hard_reg_limit      = 16384,
```

Comment:
we discussed this a bit on the telecon, having large numbers of xpmem attached regions make for performance problems with any operations in the kernel involving the process address space where iterations over VMAs is required.  And the attach is only done from the receiver side in the rendezvous path, so there's no chance that we could deadlock due to no stale entries in the cache being available for eviction.